### DISCUSSION TITLE
engine: cannot enter cgroupv2

### DISCUSSION THREAD

**Discussion by keax** [dagger/dagger#12398] [Help]
### What is the issue?

I encounter errors when I run Kubernetes engine (k3s) inside a Dagger pipeline.

It works well with a previous version of Dagger (0.6.4).
It seems to be related to this issue : #5593 and with a potential fix [https://github.com/moby/buildkit/pull/4308](https://github.com/moby/buildkit/pull/4308)
I don't understand very well what this fix does.
Dagger version 0.9.5 uses the buildkit version [github.com/moby/buildkit](http://github.com/moby/buildkit) v0.13.0-beta1.0.20231107172747-c4f191410a41
So it is supposed to be fixed but I think I missed something.

Does anyone can explain me where is the problem ?

### Dagger version

dagger v0.9.5

### Steps to reproduce

Steps are very well explained in #5292 and work well with Dagger version 0.6.4.

### Log output

```
333: [8.71s] W0105 12:44:19.019252 17 watcher.go:93] Error while processing event ("/sys/fs/cgroup/kubepods": 0x40000100 == IN_CREATE|IN_ISDIR): readdirent /sys/fs/cgroup/kubepods: no such file or directory
333: [8.71s] E0105 12:44:19.019265 17 node_container_manager_linux.go:61] "Failed to create cgroup" err="cannot enter cgroupv2 \"/sys/fs/cgroup/kubepods\" with domain controllers -- it is in an invalid state" cgroupName=[kubepods]
333: [8.71s] E0105 12:44:19.019279 17 kubelet.go:1431] "Failed to start ContainerManager" err="cannot enter cgroupv2 \"/sys/fs/cgroup/kubepods\" with domain controllers -- it is in an invalid state"
```

**Comment by jedevc** [comment]
@keax the most recent release of dagger (v0.9.7) uses a version of buildkit that has the fix you've linked in https://github.com/moby/buildkit/pull/4308. You can upgrade to get the fix - there's not really a way of fixing it without that - though you could upgrade your underlying operating system to a kernel version that supports cgroups v2 as well.

The issue is that buildkit introduced an issue which attempted to enable cgroup namespaces when they're supported - however, as a mistake they were being added to cgroupsv1, where they aren't supported. The fix ensures that they are not mistakenly enabled.

**Comment by keax** [comment]
(I don't known if I have to open a new issue, or re-open this one)
@jedevc I tried to upgrade to 0.9.7 and the problem still persists :

```
/ # dagger version
dagger v0.9.7 (registry.dagger.io/engine) linux/amd64
```

And it returns the same issue : 
```
cannot enter cgroupv2
```
I also checked my kernel version and cgroups version as you recommended :
```
~$ mount -l | grep cgroup
cgroup2 on /sys/fs/cgroup type cgroup2 (rw,nosuid,nodev,noexec,relatime)
```

Any idea ?

**Comment by demoaccchallenge** [comment]
Facing the same issue both on colima and on a amd64 box.

**Comment by jedevc** [comment]
Hm, clearly this is not resolved by the buildkit upgrade, reopening.

**Comment by marcosnils** [comment]
:wave: found what's happening as it's related to cgroup nesting not being properly set. More info here: https://discord.com/channels/707636530424053791/1203019407656353842/1203166561750220830

also updated the snippet in the original issue here https://github.com/dagger/dagger/issues/5292#issuecomment-1593750070 to reflect the changes.
